depends: maintain glibc backcompat on focal for x86_64-apple-darwin11

xanimo opened the pull request:
In preparation for `Ubuntu Bionic`'s end of life slated in April of 2023 which in turn has prompted `GH Actions Bionic` runner becoming fully deprecated on December 1st, 2022 (#3095), this PR kicks off the migration process by bumping `x86_64-apple-darwin11`'s CI runner OS to `Ubuntu Focal 20.04`.

In order to accomplish that, I needed a minimal amount of cherry-picks from Bitcoin Core that:
 - fixed a bug that had to do with archive extraction when building depends: fff1711 
 - bumped `clang+llvm` from `3.7.1` to `6.0.1` and corresponding build tools (`native_cctools`, `ld64`, `libtapi`): 901eaa3 and 0fff9eb

Once that was pulled in I encountered an issue with QT which is fixed with the patch found in 4314966:
- After updating `clang+llvm`, `QT 5.7.1`'s qfontengine_coretext.mm files variable QFixed errors out as it's a constructor rather than a type in this context. This patch fixes that error.

After I was able to emulate a successful build inside an `ubuntu:focal` container and in the following commit 9eb81d3 that bumped `x86_64-apple-darwin11` matrix OS in CI I had to:
- add 2 architecture specific packages: `libbz2-dev`, `python-is-python3`
- drop duplicate of `libtinfo5` from `x86_64-apple-darwin11` packages

And finally I made the same changes as the previous commit only this time applicable for our gitian-osx.yml descriptor c930d8a: 
- temporarily bump docker conditional in gitian-builder to build bin/make-base-vm with ubuntu focal in gitian-build.sh script
- bump suite to focal in gitian-osx.yml and replace `python` and `python-setuptools` with `python3`
- additionally add `python-is-python3` for PATH purposes and add `libtinfo5` and `xorriso`
- alphabetize packages list for easier management

Note that this currently disallows a seamless gitian build and folks will have to (after failure when transitioning from osx to windows), build a base-vm for bionic in order to resume building windows and linux archives. I've not yet completed that step but did successfully compile for osx:
```
cf3eeab11481f3b60e4e2791dc29a11b099c025b9d490d8d4b68e4776e40b868  dogecoin-1.14.7-osx-unsigned.dmg
37d854c9fed8541dca877830338ba72bb4a89b229d7a71e16414b535205b2582  dogecoin-1.14.7-osx-unsigned.tar.gz
cf069aeec64d6b848fb0e4a18e6dab03111d33512acea5a027c7bc43942bc30d  dogecoin-1.14.7-osx64.tar.gz
7136259be06e9ce6bd791476921197ec6087d87461b169c6a429b2869077a156  src/dogecoin-1.14.7.tar.gz
d96a40c1e43924ac54e36eb71a272fddef7c00e807c14e8b9920ec4a8f726175  dogecoin-osx-1.14-res.yml
```

patricklodder commented:
First off, thanks a bunch for this! The osx-on-focal build problems have been haunting us for over a year and this seems to be a working solution. Thank you!

I did a check on (what is now) 4314966 before this PR was raised, and according to `otool`, backward compatibility in terms of dynamically linked libraries is exactly the same as with the 1.14.6 released binary (which was built with clang 3.7, so I think we ought to be thankful for clang being awesome too 😁). The gist with output can be found [here](https://gist.github.com/patricklodder/1aa514e6e1ffaeaaaee9608253282679) and I'll see if I can update the symbol check script in `contrib/devtools` to use `otool` on cross-compiles and automate this process for CI, as Bitcoin also does in newer versions of the script.

xanimo commented:
It's my pleasure and I forgot to note that those osx checksums were from a build reproducible (fingers crossed) with the following invocation (with assumption you do so in ./contrib so adjust accordingly): `./gitian-build.sh --docker --setup --build -u https://github.com/xanimo/dogecoin -c c930d8a66861e46a5e2d28e2055bc4a57c022410`

patricklodder commented:
Good news: My gitian+docker built shasums match yours
Bad news: I am still working on getting a reliable gitian+lxc setup for this

Some high-level observations:

- I think that we'll need to separate fff1711 and 4314966 from this PRs (pref individually.) These influence every build, not just gitian-osx, so we'll need to validate those changes across all builds, which is difficult when we have different base OSs. I assume that these changes are also valid with the current bionic-based gitian descriptors - if I am wrong about that ignore this bullet point. 
- Can we instead of `python-is-python3` just call `python3` instead of `python` like we do everywhere else? What's causing it to fail without this package?
- "alphabetize packages list for easier management" and "drop duplicate of libtinfo5 from x86_64-apple-darwin11 packages"- can we do these separately? It's hard to judge the diff this way (now and in the future). I'm okay with a refactor commit in a preceding PR (much like the commits from bullet point 1)
- "temporarily bump docker conditional in gitian-builder to build bin/make-base-vm with ubuntu focal in gitian-build.sh" script -> please put this in a separate commit

xanimo commented:
> Good news: My gitian+docker built shasums match yours Bad news: I am still working on getting a reliable gitian+lxc setup for this

I've just built `./gitian-build.sh --docker --setup --build -u https://github.com/xanimo/dogecoin -c a6bcaedcea231bfe4dced1867b4c30824e95353a` and got the following checksums for osx:

```
39244aac1338b4ff91fb53f00204b4516e04a9379c235fc544e26c892f50bf09  gitian-output/dogecoin-binaries/a6bcaedcea231bfe4dced1867b4c30824e95353a/dogecoin-1.14.7-osx64.tar.gz
4101e65ae2d24600494c61db745fea23197d1019dce2285532befcfbd4007e69  gitian-output/dogecoin-binaries/a6bcaedcea231bfe4dced1867b4c30824e95353a/dogecoin-1.14.7-osx-unsigned.dmg
f8cc436df77d5d88ff606ef4760a1423ad7507a416dd0809d2a800baa0dca320  gitian-output/dogecoin-binaries/a6bcaedcea231bfe4dced1867b4c30824e95353a/dogecoin-1.14.7-osx-unsigned.tar.gz
1d2e8e831c9492ee630e558ba0d248a4807b489b7249f5a09c603edaa7dee83c  gitian-output/dogecoin-binaries/a6bcaedcea231bfe4dced1867b4c30824e95353a/dogecoin-1.14.7.tar.gz
```

I made a note to modify the make-base-vm invocation for everything (not just docker) in the gitian-build script so hopefully that will work on yours and others systems.

> Some high-level observations:
> 
>     * I think that we'll need to separate [fff1711](https://github.com/dogecoin/dogecoin/commit/fff171115c37dccd8b48d927c1c46e6e16c05cae) and [4314966](https://github.com/dogecoin/dogecoin/commit/4314966ff82efa0a30c00ee7ed1ee1262fc2ded1) from this PRs (pref individually.) These influence every build, not just gitian-osx, so we'll need to validate those changes across all builds, which is difficult when we have different base OSs. I assume that these changes are also valid with the current bionic-based gitian descriptors - if I am wrong about that ignore this bullet point.

It turns out fff1711 is not entirely necessary so I'm holding back on opening an additional PR for now but I have dropped it from this one. I did however open one for 4314966 which can be found here: #3141

>     * Can we instead of `python-is-python3` just call `python3` instead of `python` like we do everywhere else? What's causing it to fail without this package?

I've stricken `python-is-python3` and instead updated any use of `python` to `python3` in `native_biplist.mk`, `native_ds_store.mk` and `native_mac_alias.mk` in `dogecoin/depends/packages/` (These packages were points of failure when building depends on ubuntu >= `focal`)

>     * "alphabetize packages list for easier management" and "drop duplicate of libtinfo5 from x86_64-apple-darwin11 packages"- can we do these separately? It's hard to judge the diff this way (now and in the future). I'm okay with a refactor commit in a preceding PR (much like the commits from bullet point 1)

I have split these up into more atomic commits per your request.

>     * "temporarily bump docker conditional in gitian-builder to build bin/make-base-vm with ubuntu focal in gitian-build.sh" script -> please put this in a separate commit

Done :+1:

patricklodder commented:
> It turns out https://github.com/dogecoin/dogecoin/commit/fff171115c37dccd8b48d927c1c46e6e16c05cae is not entirely necessary so I'm holding back on opening an additional PR for now but I have dropped it from this one.

Yes, we can either do it up front or after we have a full gitian/ci upgrade to focal merged. If it doesn't hurt in bionic, feel free to open it up now - from the commit description I feel it makes building depends more portable

> I did however open one for https://github.com/dogecoin/dogecoin/commit/4314966ff82efa0a30c00ee7ed1ee1262fc2ded1 which can be found here: https://github.com/dogecoin/dogecoin/pull/3141

I was anyway planning to tag this PR with "do not merge yet" so I'm doing that and am not going to map out cross-PR dependencies for this in the review board right now, because updating the build environment is very much an all-or-nothing thing.

> I have split these up into more atomic commits per your request.

Can you pull these prior and separate like with #3141? I am trying to separate independent refactor / cleanup from functional change - the whole of switching to focal is going to be huge so I think it's in our best interest to keep the scope of the PRs as small as possible.

patricklodder commented:
Tagged do-not-merge until we have all gitian build targets ready for merge.

xanimo commented:
Absolutely, I'll leave this intact as a reference point for what was needed to achieve successful build on CI and macOS gitian build and pull the ones that are compatible with our existing systems into standalone PRs.

Re fff1711, I realized late last night/early this morning that [3773c63](https://github.com/dogecoin/dogecoin/pull/3133/commits/3773c63d6e125adae928d31dcb28491e9809f2dc) and [0e3b006](https://github.com/dogecoin/dogecoin/pull/3133/commits/0e3b0064c930ab4cee7325f19779da5fb3fb1822) already include partial changes related (`tar --no-same-owner...`) so I think it would behoove us to pull that which I'll do in a separate standalone PR akin to #3141. And agreed on the depends build portability aspect. I can confirm that it's unobtrusive and works on bionic too according to this run from yesterday: https://github.com/xanimo/dogecoin/actions/runs/3260954920

No worries re do not merge, this is a big change so best to dot our i's and cross our t's :+1:

patricklodder commented:
Tagged Blocked in review board, until we merge the extracted PRs, rebase and have a full focal gitian for all build targets
